File: 42-trapping-rain-water/solution.cpp
// Runtime: 11ms, Beats 66.15% of users with C++
// Memory: 23.33MB, Beats 5.02% of users with C++

class Solution {
public:
    int trap(vector<int>& height) {
        int n=height.size();
        vector<int> maxLeft(n);
        vector<int> maxRight(n);
        int water=0,maxi=0;
        for(int i=0;i<n;i++)
        {
            maxLeft[i]=maxi;
            maxi=max(height[i],maxi);
        }
        maxi=0;
        for(int i=n-1;i>=0;i--)
        {
            
            maxRight[i]=maxi;
            maxi=max(height[i],maxi);
        }
        int quantity;
        for(int i=0;i<n;i++)
        {
            quantity=min(maxLeft[i],maxRight[i]);
            water+=max(0,quantity-height[i]);
        }
        return water;
    }
};

File: 42-trapping-rain-water/solution.py
# Runtime: 117ms, Beats 38.46% of users with Python3
# Memory: 18.54MB, Beats 67.10% of users with Python3

class Solution:
    def trap(self, height: List[int]) -> int:
        n=len(height)
        maxLeft=[0]*n
        maxRight=[0]*n
        water=0
        maxi=0
        quantity=0
        for i in range(n):
            maxLeft[i]=maxi
            maxi=max(height[i],maxi)
        maxi=0
        for i in range(n-1,-1,-1):
            maxRight[i]=maxi
            maxi=max(height[i],maxi)
        for i in range(n):
            quantity=min(maxLeft[i],maxRight[i])-height[i]
            water+=max(quantity,0)
        return water
